File: Jump_to_Algorithm2/week32/KyoungyeonJu/삼각달팽이.py
def solution(n):
    answer = [[0]*i for i in range(1, n+1)]
    rot = [[1, 0], [0, 1], [-1, -1]]  # 반시계 방향으로 채우기: r+1 증가, c+1 증가, r-1감소와 c-1감소 반복
    num, r, c = 1, -1, 0
    rot_num = 0

    for i in range(n, 0, -1):
        rr, cc = rot[rot_num % 3]
        for j in range(i):
            nr, nc = r + rr, c + cc
            answer[nr][nc] = num
            r, c = nr, nc
            num += 1
        rot_num += 1

        # 2차원 리스트 -> 1차원 리스트
        real_answer = []
        while len(answer):
            list = answer.pop(0)
            real_answer.extend(list)

        return real_answer


"""
# 2차원 리스트 -> 1차원 리스트
1. pop하고 extend
테스트 7 〉	통과 (120.87ms, 43MB)
테스트 8 〉	통과 (107.37ms, 43.2MB)
테스트 9 〉	통과 (105.76ms, 43.3MB)

2. itertools.chain(*list)
테스트 7 〉	통과 (154.51ms, 43.1MB)
테스트 8 〉	통과 (142.49ms, 43.2MB)
테스트 9 〉	통과 (151.86ms, 43.2MB)

3. sum(answer, [])
테스트 7 〉	통과 (1723.80ms, 45.3MB)
테스트 8 〉	통과 (1806.23ms, 45.2MB)
테스트 9 〉	통과 (1695.21ms, 45.2MB)
"""

File: Jump_to_Algorithm2/week32/dayeongYoo/삼각 달팽이.java
class Solution {
    public int[] solution(int n) {
        // 세 방향 존재 -> 아래, 오른쪽, 대각선(x,y 둘다 감소)
        int x = -1; // 아래 이동은 x 좌표 감소
        int y = 0; // 오른쪽 이동은 y 좌표 증가
        int num = 1; // 삼각 달팽이 채우기 값
        // 삼각형
        int[][] tri = new int[n][n];
        // 삼각형의 크기 ( 1 ~ n 까지 합)
        int[] answer = new int[n * (n + 1) / 2];

        for (int i = 0; i < n; i++) {
            for (int j = i; j < n; j++) {
                if (i % 3 == 0) { // 아래
                    x++;
                } else if (i % 3 == 1) { // 오른쪽
                    y++;
                } else if (i % 3 == 2) { // 대각선
                    x--;
                    y--;
                }
                tri[x][y] = num++;
            }
        }
        // 결과 배열에 담기
        int index = 0;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (tri[i][j] == 0) break;
                answer[index++] = tri[i][j];
            }
        }
        return answer;
    }
}